draft-ietf-quic-tls.md

```diff
@@ -866,6 +867,81 @@ the connection that is hosted on stream 0.  This sequence number is not visible
 to QUIC.
 
 
+## Packet Number Protection {#pn-encrypt}
+
+QUIC packets are protected using a key that is derived from the current set of
+secrets.  The key derived using the "pn" label is used to protect the packet
+number from casual observation.  The packet number protection algorithm depends
+on the negotiated AEAD.
+
+Packet number protection is applied after packet protection is applied (see
+{{aead}}).  The ciphertext of the packet is sampled and used as input to an
+encryption algorithm.  For packets with a long header, the ciphertext starting
+immediately after the packet number is used (that is, octet 17 onwards).  For
+packets with a short header, the packet number length is not known, so it is
+assumed to be its largest possible length (4 octets).  Thus, for a short header,
+the sampled ciphertext starts at either octet 5 when the connection ID is
+omitted, or octet 13 when the connection is present.
+
+The protected packet might not have produced enough input for the negotiated
+packet protection algorithm.  This might happen if a packet with a short header
+contains minimal data and uses a packet number encoding that is shorter than 4
+octets.  Additional zero octets are added to the end of the sequence to reach
+the required amount of data.
+
+Before a TLS ciphersuite can be used with QUIC, a packet protection algorithm
+MUST be specifed for the AEAD used with that ciphersuite.  This document defines
+algorithms for AEAD_AES_128_GCM, AEAD_AES_128_CCM, AEAD_AES_256_GCM,
+AEAD_AES_256_CCM (all AES AEADs are defined in {{!RFC5116}}), and
+AEAD_CHACHA20_POLY1305 ({{!CHACHA=RFC7539}}).
+
```

Comment:
I find this whole section hard to understand and hard to implement. I would like to see much more uniformity between the various constructs. I get the following:

1) We are going to use the same encryption algorithm for PN as the encryption algorithm part of AEAD.

2) There will be a specific key for PN encryption, different from the key used for data encryption. 

3) The PN encryption will use a 16 byte sample of the packet's ciphertext as nonce.

4) Our AEAD specification tells us that the nonce is obtained by combining a preset IV with a sequence number. In our case, we will use the first 4 bytes of the sample as a 32 bit sequence number, and the last 12 bytes of the sample as IV, zero padded to the length required by the selected algorithm if necessary. 

5) Conceptually, AEAD algorithm encode an N byte filed by XOR it with a crypto-graphically generated stream of data, plus a 16 byte MAC. When doing PN encryption, we only use the first N bytes of the output, and do not compute or check the MAC. That is, PN encryption uses the "XOR" subset of the AEAD algorithm.

6) Using only the XOR part of the AEAD algorithm requires specific optimizations depending on the specific AEAD algorithm. These optimizations are presented in the following sections for AES and CHACHA20.

7) One way to test that a PN encryption implementation is correct is to compute the AEAD encryption of the PN with the selected IV and sequence number, and verify that the AEAD algorithm provides for the same initial bytes as the PN encryption implementation.

Do we agree on that?

Reply:
You can consider this an application of the AEAD, but I think that would be an error.  I prefer to think of this as using the same primitives that the AEAD uses.  That is, we build AEAD_AES_128_GCM using AES, and so we use AES here.

There are commonalities in the design, and I suspect that if you abuse your AEAD interface hard enough, you can get the AEAD to produce the right bits, but that's probably inefficient (you don't need the MAC, but the AEAD will calculate one) and not a general solution (there is no guarantee that the AEAD puts the MAC at the end, or doesn't do other post-processing that would might this design).

The reason for that is part of what we sample from the ciphertext of the packet is fed into the counter part and most AEADs don't let you tweak that (all AEADs I've seen have an internal counter that starts from 0 for each invocation).  We need that counter because the 96-bit nonce doesn't have enough entropy.  128 bits ensures that we don't have to worry about collisions, because we can only send 2^62 packets.

So 1 - shared primitives, not same, 2 - agreed, 3 - this will depend on AEAD, unfortunately, 4 - that's internal AEAD details, 5 - that's AEAD-specific, 6 - see point about shared primitives, 7 - this might be true for the current AEADs if you can abuse them enough, but it's not generally true.

Reply:
I think Kazuho has a plausible generalization with his CTR mode API. Maybe we should find a way to document that.

Reply:
We can't know that there is a CTR mode for the underlying primitive.  I'm happy to include some advice along the lines of "for AES, this is just CTR mode", or a hint that in general CTR mode is a good design, but any further generalization is a presumption that might make it harder to deploy a new cipher.

Reply:
I can understand @martinthomson's point that if a stream cipher (or block cipher in CTR mode) exists beneath an AEAD is an implementation detail. Hence we need a mapping.

I also agree with @huitema that it would help implementors if the AES algorithm was explained as CTR mode. That would help people come up with an implementation that shares the code between the cipher suites.

Reply:
With the updated text, I understand very well where to pick the sample. That's good. But there is still an issue. You note that "For packets with a short header, the packet number length is not known before decryption". If that is the case, then the receiver has to decode the first PN byte, get the length, then decode the other bytes. That means either two passes of encryption, or doing the XOR manually. Not great.

I would prefer "encrypt from the beginning of the PN to just before the start of the sample". One pass, no fuss.

Reply:
Why not copy out 4 octets, XOR those, and then read off the PN?  Any excess will be nonsense, but you aren't using those bits.  That should be perfectly efficient and it's very easy to implement.

Reply:
The CCM mode has a 16 byte header that includes the nonce - it won't be encrypted unless it is taken care of explicitly for that mode and if it isn't encrypted, encrypting the packer number is pointless.

Reply:
@mikkelfj: I believe you have misunderstood how CCM is used in TLS (and presumably in QUIC).

The inputs to CCM encryption are:

- A nonce N
- The key
- The plaintext
- The associated data A

The inputs to CCM decryption are:

- A nonce N
- The key
- The ciphertext
- The associated data A

In some versions of CCM, you explicitly render N on the wire, but in TLS 1.3, you would not do this, but rather (as in AES-GCM), derive it from the PN. There is then no need to render the nonce on the wire, and it behaves as GCM does.

Reply:
That is reassuring to hear. I original found an older RFC 3610, but I then checked  TLS 1.3 draft 23 which referred to RFC 5116 which in turn referred via https://tools.ietf.org/html/rfc5116#ref-CCM to NIST https://csrc.nist.gov/publications/detail/sp/800-38c/final appendix A, which lists said header with NONCE - but it wasn't very clearly formulated, so I (apparently incorrectly) assumed it was similar to https://tools.ietf.org/html/rfc3610 except for fixed parameters chosen in the header (as I would have expected).

Reply:
Is this resolved now? Does the text below address @huitema and @kazuho 's concerns?

Reply:
I believe so.